src/test/rpc/ShardArchiveHandler_test.cpp

```diff
@@ -0,0 +1,378 @@
+                env.app(),
+                env.app().getJobQueue(),
+                false);
+            BEAST_EXPECT(handler);
+
+            auto initialized = handler->init();
+            BEAST_EXPECT(initialized);
+
+            auto server = createServer(env);
+            auto host = server.local_endpoint().address().to_string();
+            auto port = std::to_string(server.local_endpoint().port());
+            server.stop();
+
+            Downloads dl;
+
+            for(int i = 1; i <= 10; ++i)
+            {
+                dl.push_back({i,
+                    (boost::format("https://%s:%d/%d.tar.lz4") %
+                         host % port % i).str()});
+            }
+
+            for(auto const & entry : dl)
+            {
+                parsedURL url;
+                parseUrl(url, entry.second);
+                handler->add(entry.first, {url, entry.second});
+            }
+
+            checkState(handler);
+
+            auto stateDir = RPC::ShardArchiveHandler::getDownloadDirectory(
+                env.app().config());
+
+            boost::filesystem::copy_file(stateDir / stateDBName,
+                boost::filesystem::path(tempDir.path()) / stateDBName);
+
+            auto result = handler->start();
+            BEAST_EXPECT(result);
+
+            std::unique_lock<std::mutex> lock(handler->m_);
+            while(!handler->archives_.empty())
+            {
+                lock.unlock();
+                std::this_thread::sleep_for(std::chrono::milliseconds(100));
+                lock.lock();
+            }
+
+            BEAST_EXPECT(!boost::filesystem::exists(stateDir));
+
+            boost::filesystem::create_directory(stateDir);
+
+            boost::filesystem::copy_file(
+                boost::filesystem::path(tempDir.path()) / stateDBName,
+                stateDir / stateDBName);
+        }
+
+        // Destroy the singleton so we start fresh in
+        // the new scope.
+        RPC::ShardArchiveHandler::instance_.reset();
+
+        auto c = jtx::envconfig();
+        auto & section = c->section(ConfigSection::shardDatabase());
+        section.set("path", tempDir.path());
+        section.set("max_size_gb", "100");
+        c->setupControl(true, true, true);
+
+        jtx::Env env(*this, std::move(c));
+
+        while(!RPC::ShardArchiveHandler::hasInstance())
+            std::this_thread::sleep_for(std::chrono::milliseconds(100));
+
+        BEAST_EXPECT(RPC::ShardArchiveHandler::hasInstance());
+
+        auto handler = RPC::ShardArchiveHandler::getInstance();
+
+        std::unique_lock<std::mutex> lock(handler->m_);
+        while(!handler->archives_.empty())
+        {
+            lock.unlock();
+            std::this_thread::sleep_for(std::chrono::milliseconds(100));
+            lock.lock();
+        }
+
+        auto stateDir = RPC::ShardArchiveHandler::getDownloadDirectory(
+            env.app().config());
+
+        BEAST_EXPECT(!boost::filesystem::exists(stateDir));
```

Comment:
Again, I would suggest checking that the folder exists before the loop starts.
```
            std::unique_lock<std::mutex> lock(handler->m_);
            BEAST_EXPECT(boost::filesystem::exists(stateDir) || handler->archives_.empty());
            while(!handler->archives_.empty())
[...]
```

Reply:
Done